File: src/main/java/ski/mashiro/data/CourseData.java
package ski.mashiro.data;

import ski.mashiro.constant.StatusCodeConstants;
import ski.mashiro.dto.CourseDto;
import ski.mashiro.dto.CourseSearchDto;
import ski.mashiro.dto.Result;
import ski.mashiro.file.UserFile;
import ski.mashiro.pojo.User;
import ski.mashiro.util.CourseUtil;
import ski.mashiro.util.FormatterUtil;
import ski.mashiro.util.HttpUtils;

import java.time.LocalDate;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Queue;

/**
 * @author MashiroT
 */
public class CourseData {

    public static final Map<Long, Queue<CourseDto>> USER_TODAY_COURSE = new HashMap<>(UserFile.USERS.size());

    public static List<CourseDto> getCourse(CourseSearchDto searchDto) {
        Result<List<CourseDto>> result = HttpUtils.sendCourseReq(searchDto);
        if (result.getCode() == StatusCodeConstants.COURSE_LIST_FAILED) {
            return null;
        }
        return FormatterUtil.sortByTime(result.getData());
    }

    public static boolean refreshQueue(long qq) {
        User currUser = UserFile.USERS_MAP.get(qq);
        CourseSearchDto searchDto = new CourseSearchDto(currUser.getUid(), FormatterUtil.INT_DAY_TO_STR_MAP.get(LocalDate.now().getDayOfWeek().getValue()), currUser.getTermStartDate());
        List<CourseDto> courseDtoList = CourseData.getCourse(searchDto);
        if (courseDtoList == null) {
            return false;
        }
        CourseData.USER_TODAY_COURSE.get(qq).addAll(courseDtoList);
        while (!CourseData.USER_TODAY_COURSE.get(qq).isEmpty() && CourseUtil.isCourseExpired(CourseData.USER_TODAY_COURSE.get(qq).peek())) {
            CourseData.USER_TODAY_COURSE.get(qq).poll();
        }
        return true;
    }
}

File: src/main/kotlin/ski/mashiro/data/CourseData.kt
package ski.mashiro.data

import ski.mashiro.CourseScheduleQQAdvice
import ski.mashiro.constant.Consts
import ski.mashiro.entity.Course
import ski.mashiro.entity.User
import ski.mashiro.service.impl.CourseServiceImpl
import ski.mashiro.util.TimeUtils
import java.time.LocalDateTime
import java.util.*

object CourseData {

    private val logger = CourseScheduleQQAdvice.logger

    val courseMap = HashMap<Long, Queue<Course>>(10)

    fun getCourseData(now: LocalDateTime, user: User) {
        val result = CourseServiceImpl.getTodayCoursesFromServer(user)
        if (result.code != Consts.COURSE_LIST_SUCCESS) {
            logger.warning("用户: ${user.username} 获取今日课程失败")
            return
        }
        courseMap[user.qq] = LinkedList(
            result.data!!
                .filter {
                    now.isAfter(TimeUtils.getCourseEndTime(now, it.time))
                }.sortedWith { o1, o2 ->
                    TimeUtils.compareStrTime(now, o1.time, o2.time)
                }
        )
    }

}